checkers/dupOption_checker.go

```diff
@@ -0,0 +1,113 @@
+package checkers
+
+import (
+	"go/ast"
+	"go/token"
+	"go/types"
+
+	"github.com/go-critic/go-critic/checkers/internal/astwalk"
+	"github.com/go-critic/go-critic/linter"
+	"github.com/go-toolsmith/astcast"
+	"github.com/go-toolsmith/astfmt"
+)
+
+func init() {
+	var info linter.CheckerInfo
+	info.Name = "dupOption"
+	info.Tags = []string{linter.DiagnosticTag, linter.ExperimentalTag}
+	info.Summary = "Detects duplicated option func arg"
+	info.Before = `doSomething(name,
+		withWidth(w),
+		withHeight(h),
+		withWidth(w),
+)`
+	info.After = `doSomething(name,
+		withWidth(w),
+		withHeight(h),
+)`
+
+	collection.AddChecker(&info, func(ctx *linter.CheckerContext) (linter.FileWalker, error) {
+		c := &dupOptionChecker{ctx: ctx}
+		return astwalk.WalkerForExpr(c), nil
+	})
+}
+
+type dupOptionChecker struct {
+	astwalk.WalkHandler
+	ctx *linter.CheckerContext
+}
+
+func (c *dupOptionChecker) VisitExpr(expr ast.Expr) {
+	call := astcast.ToCallExpr(expr)
+	variadicArgs, argType := c.getVariadicArgs(call)
+	if len(variadicArgs) == 0 {
+		return
+	}
+
+	if !c.isOptionType(argType) {
+		return
+	}
+
+	dupArgs := c.findDupArgs(variadicArgs)
+	for _, arg := range dupArgs {
+		c.warn(arg)
+	}
+}
+
+func (c *dupOptionChecker) getVariadicArgs(call *ast.CallExpr) ([]ast.Expr, types.Type) {
+	if len(call.Args) == 0 {
+		return nil, nil
+	}
+
+	// skip for someFunc(a, b ...)
+	if call.Ellipsis != token.NoPos {
+		return nil, nil
+	}
+
+	funType := c.ctx.TypeOf(call.Fun)
+	sign, ok := funType.(*types.Signature)
+	if !ok || !sign.Variadic() {
+		return nil, nil
+	}
+
+	last := sign.Params().Len() - 1
+	sliceType, ok := sign.Params().At(last).Type().(*types.Slice)
+	if !ok {
+		return nil, nil
+	}
+	argType := sliceType.Elem()
+	return call.Args[last:], argType
+}
+
+func (c *dupOptionChecker) isOptionType(typeInfo types.Type) bool {
+	typeInfo = typeInfo.Underlying()
+
+	sign, ok := typeInfo.(*types.Signature)
+	if !ok {
+		return false
+	}
+
+	if sign.Params().Len() == 0 {
```

Comment:
Not sure about this one. Looks like the 2nd call should be reported:

```
	_ = newPanel("hello",
		withAutosize(),
		withAutosize(),
	)
```

Reply:
it check the type of return result from `withAutoSize()` , not `withAutoSize`'s signature, I add more test case like

`withDefaultWidth()` and `defaultWidth` option, how to create the option is ignored, only check the aftfmt output (code) is same.

Reply:
Well, `autoSize` can set some field to `true` implicitly. So, param isn't required.

Also, `withDefaultConfig()` which sets bunch of fields at once.

Reply:
yes, you are right, if some Option is the type `func() Config`, it can has no params and set some fields, 

but i don't think this is a good `Option`, like this code, `Option1`, `Option2`, `Option3` is not support,  

how do you think? 🤔

```go
type Config struct {
	width  int
	height int
}

type Option1 func()
type Option2 func() error
type Option3 func() (Config, error)
type Option4 func(c *Config)
type Option5 func(c *Config) error
type Option6 func(c Config) Config
```

Reply:
I think that's fine for now. Let's see how it will work in the wild.